stan/math/rev/mat/functor/algebra_solver_fp.hpp

```diff
@@ -0,0 +1,342 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_FP_SOLVER_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_FP_SOLVER_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/mat/fun/to_array_1d.hpp>
+#include <stan/math/prim/mat/fun/to_vector.hpp>
+#include <stan/math/prim/mat/fun/value_of.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/prim/mat/fun/mdivide_left.hpp>
+#include <stan/math/prim/scal/err/check_nonnegative.hpp>
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/rev/mat/functor/kinsol_data.hpp>
+#include <stan/math/prim/mat/err/check_flag_sundials.hpp>
+
+#include <kinsol/kinsol.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <nvector/nvector_serial.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <algorithm>
+
+namespace stan {
+namespace math {
+
+/**
+ * KINSOL algebraic system data holder.
+ *
+ * @tparam F1 functor type for system function.
+ * @tparam F2 functor type for jacobian function. Default is 0.
+ *         If 0, use rev mode autodiff to compute the Jacobian.
+ */
+template <typename F>
+struct KinsolFixedPointEnv {
+  const F& f_;
+  const Eigen::VectorXd y_dummy;
+  const Eigen::VectorXd& y_;
+  const size_t N_;
+  const size_t M_;
+  const std::vector<double>& x_r_;
+  const std::vector<int>& x_i_;
+  std::ostream* msgs_;
+
+  void* mem_;
+  N_Vector nv_x_;
+  N_Vector nv_u_scal_;
+  N_Vector nv_f_scal_;
```

Comment:
Just creepin:

Does sundials expose all of these things at the global scope? It may not be terrible to just wrap those includes into a namespace so we don't have to worry about exposing user space to all of the sundials stuff. If that's easy it would be nice but if that's hard then it's whateva'

Reply:
We should do it. 
It's not hard. 
Nobody wants to do/maintain it.

Reply:
Story of my life I get that